3_harmonize/src/clean_wqp_data.R

```diff
@@ -3,29 +3,103 @@
+#' organization, site id, date, time, characteristic name, and sample fraction. 
+#' However, these options can be customized by passing a vector of column names 
+#' to the argument `duplicate_definition`.
 #' 
 #' @returns 
 #' Returns a formatted and harmonized data frame containing data downloaded from 
 #' the Water Quality Portal, where each row represents a unique data record.
 #' 
-clean_wqp_data <- function(wqp_data, char_names_crosswalk){
+clean_wqp_data <- function(wqp_data, char_names_crosswalk,
+                           commenttext_missing = c('analysis lost', 'not analyzed', 
+                                                   'not recorded', 'not collected', 
+                                                   'no measurement taken'),
+                           duplicate_definition = c('OrganizationIdentifier',
+                                                    'MonitoringLocationIdentifier',
+                                                    'ActivityStartDate', 
+                                                    'ActivityStartTime.Time',
+                                                    'CharacteristicName', 
+                                                    'ResultSampleFractionText')){
 
   # Clean data and assign flags if applicable
   wqp_data_clean <- wqp_data %>%
     # harmonize characteristic names by assigning a common parameter name
     # to the groups of characteristics supplied in `char_names_crosswalk`.
-    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name"))
+    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name")) %>%
+    # flag true missing results, i.e. when result measure value and detection
+    # limit value are both NA, when "not reported" is found in the column
+    # "ResultDetectionConditionText", or when any of the strings from
+    # `commenttext_missing` are found in the column "ResultCommentText".
+    mutate(flag_missing_result = case_when(
+      is.na(ResultMeasureValue) & is.na(DetectionQuantitationLimitMeasure.MeasureValue) ~ TRUE,
+      grepl("not reported", ResultDetectionConditionText, ignore.case = TRUE) ~ TRUE,
+      grepl(paste(commenttext_missing, collapse = "|"), ResultCommentText, ignore.case = TRUE) ~ TRUE
+    )) %>%
+    # Flag duplicate records
+    flag_duplicates(., duplicate_definition)
   
   return(wqp_data_clean)
   
   
 }
 
+
+#' @title Flag duplicated records
+#' 
+#' @description 
+#' Function to flag duplicated rows based on a user-supplied definition
+#' of a duplicate record. Additional flags can help reconcile duplicate sets.
+#' For example, a flag "flag_duplicate_drop_random" is assigned based on 
+#' randomly selecting the first record from a duplicated set, making it 
+#' easier to drop all others from the dataset. 
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a unique data record.
```

Comment:
Since this function is flagging duplicates, I don't think it is correct to say each row is unique
```suggestion
#' where each row represents a data record.
```